Add transaction size option

jzhou77 opened the pull request:
This allows client to set transaction size as a database option, or as a transaction option.

This fixes #1681.

alecgrieser commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2411,13 +2417,16 @@ TransactionOptions::TransactionOptions(Database const& cx) {
+	uint32_t oldSizeLimit = customTransactionSizeLimit;
 	memset(this, 0, sizeof(*this));
 	maxBackoff = cx->apiVersionAtLeast(610) ? oldMaxBackoff : cx->transactionMaxBackoff;
+	customTransactionSizeLimit = oldSizeLimit;
```
This is to make the custom size limit persist across reset (like max backoff, max retries, and timeout)? I think that's probably fine (and makes sense for consistency with those options, I suppose), though we should document that.

jzhou77 replied:
Added to 'option-set-size-limit-blurb' and 'option-db-tr-size-limit-blurb' in the documentation.

alecgrieser commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2753,8 +2762,9 @@ Future<Void> Transaction::commitMutations() {
 		}
 
-		if (transactionSize > (options.customTransactionSizeLimit == 0 ? (uint64_t)CLIENT_KNOBS->TRANSACTION_SIZE_LIMIT : (uint64_t)options.customTransactionSizeLimit))
+		if (transactionSize > options.customTransactionSizeLimit) {
```
Should the parameter just be `sizeLimit` instead of `customTransactionSizeLimit`? Maybe there is a reason that's not obvious from the changed code, but I think this line looks weird now. (Like, if I just read this line, I might think, "why doesn't this handle the case where a custom size limit isn't set?".)

jzhou77 replied:
Good suggestion. I was also thinking about changing it.

alecgrieser commented on fdbclient/vexillographer/fdb.options:
```diff
@@ -145,6 +145,9 @@ description is not currently required but encouraged.
             description="Set the maximum amount of backoff delay incurred in the call to ``onError`` if the error is retryable. This sets the ``max_retry_delay`` option of each transaction created by this database. See the transaction option description for more information." />
+    <Option name="transaction_size_limit" code="503"
+			paramType="Int" paramDescription="value in bytes"
+			description="Set the maximum transaction size which, if exceeded, will cause the transaction to be cancelled. Default to 10,000,000 bytes." />
```
White space (`paramType` here should be aligned with `paramType` in `transaction_max_retry_delay`).

alecgrieser commented on fdbclient/vexillographer/fdb.options:
````diff
@@ -200,7 +203,10 @@ description is not currently required but encouraged.
+    <Option name="max_transaction_size" code="503"
+			paramType="Int" paramDescription="value in bytes"
+			description="Set the maximum transaction size which, if exceeded, will cause the transaction to be cancelled. Valid parameter values are ``[32, 10,000,000]```." />
+	<Option name="snapshot_ryw_enable" code="600"
````
White space

alecgrieser commented on fdbclient/vexillographer/fdb.options:
```diff
@@ -200,7 +203,10 @@ description is not currently required but encouraged.
             paramType="Int" paramDescription="value in milliseconds of maximum delay"
             description="Set the maximum amount of backoff delay incurred in the call to ``onError`` if the error is retryable. Defaults to 1000 ms. Valid parameter values are ``[0, INT_MAX]``. If the maximum retry delay is less than the current retry delay of the transaction, then the current retry delay will be clamped to the maximum retry delay. Prior to API version 610, like all other transac ...
-    <Option name="snapshot_ryw_enable" code="600"
+    <Option name="max_transaction_size" code="503"
```
Should this be `size_limit`? I think it's fine if the variables inside are `max_transaction_size`, but I think we want this to be `size_limit` so that the transaction option is just `setSizeLimit` and then the database option is `setTransactionSizeLimit` (which has parallelism with the other database/transaction option pairs).

alecgrieser commented on documentation/sphinx/source/api-ruby.rst:
```diff
@@ -22,13 +22,15 @@
 .. |retry-limit-transaction-option| replace:: :meth:`Transaction.options.set_retry_limit`
 .. |timeout-transaction-option| replace:: :meth:`Transaction.options.set_timeout`
 .. |max-retry-delay-transaction-option| replace:: :meth:`Transaction.options.set_max_retry_delay`
+.. |size-limit-transaction-option| replace:: :meth:`Transaction.options.set_transaction_size_limit`
```
I think this would be `set_size_limit` with the other suggested change or `set_transaction_max_size` without it.

alecgrieser commented on documentation/sphinx/source/api-python.rst:
```diff
@@ -24,13 +24,15 @@
 .. |retry-limit-transaction-option| replace:: :func:`Transaction.options.set_retry_limit`
 .. |timeout-transaction-option| replace:: :func:`Transaction.options.set_timeout`
 .. |max-retry-delay-transaction-option| replace:: :func:`Transaction.options.set_max_retry_delay`
+.. |size-limit-transaction-option| replace:: :func:`Transaction.options.set_transaction_size_limit`
```
I think this would be `set_size_limit` with the other suggested change or `set_transaction_max_size` without it.

alecgrieser commented on documentation/sphinx/source/api-ruby.rst:
```diff
@@ -22,13 +22,15 @@
 .. |retry-limit-transaction-option| replace:: :meth:`Transaction.options.set_retry_limit`
 .. |timeout-transaction-option| replace:: :meth:`Transaction.options.set_timeout`
 .. |max-retry-delay-transaction-option| replace:: :meth:`Transaction.options.set_max_retry_delay`
+.. |size-limit-transaction-option| replace:: :meth:`Transaction.options.set_transaction_size_limit`
```
Also, maybe I missed it, but are there docs here for the transaction option? I could only find it for the database option.

jzhou77 replied:
Good catch. I'll add the doc for this option.

alecgrieser commented on fdbclient/DatabaseContext.h:
```diff
@@ -157,6 +157,7 @@ class DatabaseContext : public ReferenceCounted<DatabaseContext>, public FastAll
 	double transactionTimeout;
 	int transactionMaxRetries;
 	double transactionMaxBackoff;
+	int transactionMaxBytes;
```
Or `transactionMaxSize` so that it's using "size" everywhere. Up to you.

jzhou77 commented:
I've addressed all comments and added a python test that verified the size limit is effective. The test catches a bug of not setting per transaction limit, which was fixed.

alecgrieser commented on documentation/sphinx/source/api-common.rst.inc:
```diff
@@ -393,6 +397,10 @@
 
+..  |option-set-size-limit-blurb| replace::
+
+    Set the maximum transaction size limit in bytes. This size limit will persists transaction reset.
```
```suggestion
    Set the maximum transaction size limit in bytes. The size is calculated by combining the sizes of all keys and values written or mutated, all key ranges cleared, and all read and write conflict ranges. (In other words, it includes the total size of all data included in the request to the cluster to commit the transaction.) Large transactions can cause performance problems on FoundationDB clusters, so setting this limit to a smaller value than the default can help prevent the client from accidentally degrading the cluster's performance. This value must be at least 32 and cannot be set to higher than 10,000,000, the default transaction size limit. The value set by this limit will persist across transaction resets.
```

I expanded this a little bit, if these look good to you.
